File: core/src/ionium/stage/ui/TextLabel.java
package ionium.stage.ui;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.utils.Align;
import ionium.stage.Stage;
import ionium.stage.ui.skin.Palette;

public class TextLabel extends Label {

	private LocalizationStrategy i10nStrategy = new LocalizationStrategy();
	private Color textColor = new Color(1, 1, 1, 1);
	private String localizationKey = null;
	private int textAlign = Align.center;
	private boolean wrapText = false;

	public TextLabel(Stage s, Palette palette, String textKey) {
		super(s, palette);

		localizationKey = textKey;
	}

	@Override
	public void render(SpriteBatch batch, float alpha) {
		Palette palette = getPalette();

		float textY = 0;

		if ((getTextAlign() & Align.top) == Align.top) {
			textY = getHeight() - palette.labelFont.getLineHeight();
		} else if ((getTextAlign() & Align.bottom) == Align.bottom) {
			textY = 0;
		} else if ((getTextAlign() & Align.center) == Align.center) {
			textY = getHeight() * 0.5f - palette.labelFont.getCapHeight() * 0.5f;
		}

		palette.labelFont.setColor(textColor.r, textColor.g, textColor.b, textColor.a * alpha);
		palette.labelFont.draw(batch, getText(), getX(),
				getY() + textY + palette.labelFont.getLineHeight() * 0.5f, getWidth(), textAlign,
				wrapText);
		palette.labelFont.setColor(1, 1, 1, 1);
	}

	public boolean isTextWrapped() {
		return wrapText;
	}

	public TextLabel setTextWrap(boolean wrap) {
		wrapText = wrap;

		return this;
	}

	public int getTextAlign() {
		return textAlign;
	}

	public TextLabel setTextAlign(int align) {
		textAlign = align;

		return this;
	}

	public Color getColor() {
		return textColor;
	}

	public String getLocalizationKey() {
		return localizationKey;
	}

	public TextLabel setLocalizationKey(String key) {
		localizationKey = key;

		return this;
	}

	public String getText() {
		return i10nStrategy.get(getLocalizationKey());
	}

	public TextLabel setI10NStrategy(LocalizationStrategy strat) {
		i10nStrategy = strat;

		return this;
	}

	public LocalizationStrategy getI10NStrategy() {
		return i10nStrategy;
	}

}

File: core/src/main/kotlin/io/github/chrislo27/toolboks/ui/TextLabel.kt
package io.github.chrislo27.toolboks.ui

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.utils.Align
import io.github.chrislo27.toolboks.ToolboksScreen
import io.github.chrislo27.toolboks.i18n.Localization
import io.github.chrislo27.toolboks.util.gdxutils.drawCompressed
import io.github.chrislo27.toolboks.util.gdxutils.fillRect
import io.github.chrislo27.toolboks.util.gdxutils.getTextHeight
import io.github.chrislo27.toolboks.util.gdxutils.getTextWidth


/**
 * A standard text label.
 */
open class TextLabel<S : ToolboksScreen<*, *>>
    : Label<S>, Palettable {

    constructor(palette: UIPalette, parent: UIElement<S>, stage: Stage<S>) : super(palette, parent, stage) {
        this.textAlign = Align.center
    }

    var text: String = ""
    var textAlign: Int
    var isLocalizationKey = true
    var textWrapping = true
    override var background = false
    var fontScaleMultiplier: Float = 1f
    var textColor: Color? = null

    open fun getRealText(): String =
            if (isLocalizationKey)
                Localization[text]
            else
                text
    
    open fun getFont(): BitmapFont =
            palette.font

    fun setText(text: String, align: Int = this.textAlign, wrapping: Boolean = textWrapping,
                isLocalization: Boolean = isLocalizationKey): TextLabel<S> {
        this.text = text
        this.textAlign = align
        this.textWrapping = wrapping
        isLocalizationKey = isLocalization
        return this
    }

    override fun render(screen: S, batch: SpriteBatch,
                        shapeRenderer: ShapeRenderer) {
        if (background) {
            val old = batch.packedColor
            batch.color = palette.backColor
            batch.fillRect(location.realX, location.realY, location.realWidth, location.realHeight)
            batch.packedColor = old
        }

        val labelWidth = location.realWidth
        val labelHeight = location.realHeight

        val font = getFont()
        val oldScaleX = font.scaleX
        val oldScaleY = font.scaleY
        font.data.setScale(palette.fontScale * fontScaleMultiplier)
        val text = getRealText()
        val textHeightWithWrap = font.getTextHeight(text, labelWidth, true)
        val shouldWrap = textWrapping
        val textHeight = font.getTextHeight(text, labelWidth, shouldWrap)
        val textWidth = font.getTextWidth(text, labelWidth, shouldWrap)

        val y: Float
        if ((textAlign and Align.top) == Align.top) {
            y = location.realY + location.realHeight - font.capHeight
        } else if ((textAlign and Align.bottom) == Align.bottom) {
            y = location.realY + font.capHeight + (textHeight - font.capHeight)
        } else {
            y = location.realY + location.realHeight / 2 + (textHeight) / 2
        }

        val oldColor = font.color
        font.color = if (textColor != null) textColor else palette.textColor
        if (textWrapping) {
            font.draw(batch, text, location.realX, y, labelWidth, textAlign, true)
        } else {
            font.drawCompressed(batch, text, location.realX, y, labelWidth, textAlign)
        }
        font.color = oldColor
        font.data.setScale(oldScaleX, oldScaleY)
    }

}